### PR TITLE
feat: Deterministic _cq_id

### PR DESCRIPTION
**Pull request by bbernays**
#### Summary

When making cq_id deterministic we cannot use any of the existing resolver types because we cannot guarantee that the value has been fetched by that time so in this PR I call a new function after the resource has been fully resolved in `resolveResource`

---

Use the following steps to ensure your PR is ready to be reviewed

- [ ] Read the [contribution guidelines](../blob/main/CONTRIBUTING.md) 🧑‍🎓
- [ ] Run `go fmt` to format your code 🖊
- [ ] Lint your changes via `golangci-lint run` 🚨 (install golangci-lint [here](https://golangci-lint.run/usage/install/#local-installation))
- [ ] Update or add tests 🧪
- [ ] Ensure the status checks below are successful ✅

### CODE REVIEW

File: schema/resource.go
Comment (disq) [inline schema/resource.go:106]: It's also adding an empty `_cq_id` here I think. (also `_cq_parent_id` is still here, but I now think it should be included in the hash, so no worries there)

Comment (bbernays) [reply schema/resource.go:106]: But as long as it always adds the empty object it is fine

**Review by disq — APPROVED** [review]

**Review by candiduslynx — APPROVED** [review]
This will require a follow-up PR to overwrite `_cq_id` and `_cq_parent_id` where they've been skipped.

**Comment by yevgenypats** [comment]
Looks good. Can we make it an opt-int option given we don't know the impact of this on compute at scale/real-world scenarios?

File: schema/resource.go
Comment (erezrokah) [inline schema/resource.go:99]: I think this means you'll need to create a separate function for it

Comment (erezrokah) [reply schema/resource.go:100]: I think it's a bit confusing to have a function named `CalculateUniqueValue` with a param to generate a non unique value

Comment (bbernays) [reply schema/resource.go:100]: Fair! I have updated the name to be `CalculateCQID`

File: schema/resource.go
Comment (erezrokah) [inline schema/resource.go:112]: What's the reason we need to sort the PKs? It doesn't guarantee stability

File: schema/resource.go
Comment (erezrokah) [inline schema/resource.go]: Is there a reason this couldn't be implemented as a resolver?

Comment (bbernays) [reply schema/resource.go]: When making cq_id deterministic we cannot use any of the existing resolver types because we cannot guarantee that the value has been fetched by that time so in this PR I call a new function after the resource has been fully resolved in resolveResource

File: specs/source.go
Comment (erezrokah) [inline specs/source.go:53]:
```suggestion
	DeterministicCQID bool `json:"deterministic_cq_id,omitempty"`
```

Naming is hard, but maybe be more explicit of what this means

File: schema/resource.go
Comment (bbernays) [reply schema/resource.go:112]: It is an aspect of stability... If we rearrange columns (or support users rearranging) then we should not end up with a different hashed value...

File: specs/source.go
Comment (bbernays) [reply specs/source.go]: Fixed. and changed all of the other places

File: schema/resource.go
Comment (candiduslynx) [inline schema/resource.go:106]: [nit]
We do the same `MarshalBinary` -> check `err` -> `return r.Set(CqIDColumn.Name, b)` at the end of the func.
Maybe we could extract the uuid func to return either `uuid.New()` or deterministic SHA1-based one and then just call `MarshalBinary` on the returned value?

**Review by candiduslynx — CHANGES_REQUESTED** [review]
I'm OK with the changes. However, the problems with deterministic CQ ID for append mode in PG, MS SQL & other destinations should be addressed before this could be merged, hence `requesting changes` just to block the PR merge.

**Comment by bbernays** [comment]
@candiduslynx- I don't think we need to fix those issues prior to releasing this because the default behavior hasn't changed. Users have to opt in to change the behavior.

**Review by yevgenypats — CHANGES_REQUESTED** [review]
Sorry I missed it in prior reviews but we shouldn't use here hashstructure library. We already have a type system in place so no need for reflection. Either adding `Hash` function for each of them using `string` and hashing this with `io.writer` and `hasher` otherwise we don't really control the hashing mechanism also hashstructure will be much slower.

**Review by yevgenypats — APPROVED** [review]
Looks good.

Few non-blocking comments:
1) bytes.Buffer (instead of append)
2) the current slack thread - do we want to have random uuid for things without pk in any case.

**Review by yevgenypats — APPROVED** [review]

**Pull request merged**
